src/mapping/RadialBasisFctSolver.hpp

```diff
@@ -204,5 +216,88 @@ RadialBasisFctSolver::RadialBasisFctSolver(RADIAL_BASIS_FUNCTION_T basisFunction
     _qrMatrixQ = _matrixQ.colPivHouseholderQr();
   }
 }
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::solveConservative(const Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  // TODO: Avoid temporary allocations
+  // Au is equal to the eta in our PETSc implementation
+  PRECICE_ASSERT(inputData.size() == _matrixA.rows());
+  Eigen::VectorXd Au = _matrixA.transpose() * inputData;
+  PRECICE_ASSERT(Au.size() == _matrixA.cols());
+
+  // mu in the PETSc implementation
+  Eigen::VectorXd out = _decMatrixC.solve(Au);
+
+  if (polynomial == Polynomial::SEPARATE) {
+    Eigen::VectorXd epsilon = _matrixV.transpose() * inputData;
+    PRECICE_ASSERT(epsilon.size() == _matrixV.cols());
+
+    // epsilon = Q^T * mu - epsilon (tau in the PETSc impl)
+    epsilon -= _matrixQ.transpose() * out;
+    PRECICE_ASSERT(epsilon.size() == _matrixQ.cols());
+
+    // out  = out - solveTranspose tau (sigma in the PETSc impl)
+#if EIGEN_VERSION_AT_LEAST(3, 4, 0)
+    out -= static_cast<Eigen::VectorXd>(_qrMatrixQ.transpose().solve(-epsilon));
+#else
+    // Backwards compatible version
+    Eigen::VectorXd    sigma(_matrixQ.rows());
+    const Eigen::Index nonzero_pivots = _qrMatrixQ.nonzeroPivots();
+
+    if (nonzero_pivots == 0) {
+      sigma.setZero();
+    } else {
+      Eigen::VectorXd c(_qrMatrixQ.colsPermutation().transpose() * (-epsilon));
+
+      _qrMatrixQ.matrixQR().topLeftCorner(nonzero_pivots, nonzero_pivots).template triangularView<Eigen::Upper>().transpose().conjugate().solveInPlace(c.topRows(nonzero_pivots));
+
+      sigma.topRows(nonzero_pivots) = c.topRows(nonzero_pivots);
+      sigma.bottomRows(_qrMatrixQ.rows() - nonzero_pivots).setZero();
+
+      sigma.applyOnTheLeft(_qrMatrixQ.householderQ().setLength(nonzero_pivots));
+      out -= sigma;
+    }
+#endif
+  }
+  return out;
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+Eigen::VectorXd RadialBasisFctSolver<RADIAL_BASIS_FUNCTION_T>::solveConsistent(Eigen::VectorXd &inputData, Polynomial polynomial) const
+{
+  Eigen::VectorXd res;
```

Comment:
Could we use a more expressive name please? :wink:

Reply:
Yes